### PR TITLE
Rework tests even more

### PR DESCRIPTION
**Pull request by coolo**
This is just as base for discussion. This won't fly anyway as the test suite doesn't work with javascript only solutions. But I split the scheduled jobs out (and plan to move the running jobs atop the result list, not yet done) - but I also trimmed down the result list and I would like to get feedback

screenshot

### CODE REVIEW

**Comment by coolo** [comment]
the reason I merged all the settings into the name is that with the jquery datatable you can't filter by column anyway, but use the search field and would type x86_64 to get all the x86_64.

And I wondered who would sort for architecture - I think noone does.

**Comment by aaannz** [comment]
Maybe not for architecture, but I quite often sort by build#. If the search works, how about using current column header drop down as a predefined search queries? This all-in-one style looks confusing at the first glance.

**Comment by coolo** [comment]
there is no column header drop down anymore

**Comment by coolo** [comment]
But I wonder: why do you sort by build#? Can you please explain the user story?

**Comment by aaannz** [comment]
I noticed there is no column header, but I was hoping it can be recreated under jquery as well.

Hmm.. the keyword 'sort' is not right, what I meant to say is 'filter'. My user story is when I'm eager to get new TW snapshot I filter by its build# to remove stagings and others from the view.

**Comment by coolo** [comment]
you would just type build272 in the search filter - try to find mojolicous in https://build.opensuse.org/project/show/openSUSE:Factory. you don't want to sort but to filter

**Comment by ancorgs** [comment]
Anyways, now that we are backed by a proper database and we are starting to query it more directly (thanks god), nothing prevent us from adding more filters server side like we currently do with the time interval or the relevant/current/all one. We should actually remove the text search that is next to the time interval and we could use that space to add meaningful filters like "build" or "arch".

So we filter by columns/time/state server-side and then paginate, order and perform text-based search client-side. It's more coherent.

**Comment by aaannz** [comment]
That's what I'm counting on (that the search filter will work).
The problem ends with that, in this TW example, it is easier for me to just select an entry from a list than to copy&paste build# (I don't even consider hand typing TW release :) ). That's all. So no big opposition, just my lazy use case.

In case you do not find my use case attractive enough, what about leaving individual fields as non links? Use case I want to easily select fields I'm interested in with plain double click for selection clipboard. Or at least separate them so it is easier to select them.

**Comment by coolo** [comment]
the link to the build overview is missing in my proposal, which is a problem indeed. But if your user story is filtering for a specific release build, which actually overlaps heavily with the overview page, possibly we can fix that by what ancor suggests and offer a filter box with builds to query server side.

**Comment by coolo** [comment]
One thing I wan to get rid of is the time - because it defaults to 96 hours, which can result in an empty list easily. I would rather show the last 500 jobs, no matter what timeframe and offer a filter to query for old data, like the builds.

**Comment by coolo** [comment]
hmm, what I just wonder: if the builds are the interesting part - this combo box would have to go to overview directly anyway, no?

**Comment by ancorgs** [comment]
@coolo can you write down the use cases you mentioned for Tumbleweed, SLE, stagings and all that? Maybe somebody can see a pattern on them. :-)

**Comment by coolo** [comment]
1. I want to see the latest builds of all flavors for a $DISTRI-$VERSION
    (TW makes it very easy as all flavors have the same build nr, Staging:A has 2 build nrs for 2
    flavors,  SLE has also different build nrs for Server and Desktop)

 Problem here is that if we ever stop building flavor GNOME-Live, we'd need a way to ignore specific
 flavors.

 2. I want to see (older) build results for a specific ISO/Flavor for comparision. older build results for a whole $DISTRI-$VERSION only make sense in the context of TW and I think we can gently ignore that use case.

**Comment by coolo** [comment]
I changed my mind and reintegrated scheduled - so changing my proposal

screencapture-localhost-9526-tests

**Comment by coolo** [comment]
For the discussions on build number links see my idea on https://progress.opensuse.org/issues/6372

**Comment by sysrich** [comment]
I like the current proposal as an 'activity screen' - with my openQA hat on, the split between Running, Scheduled, and Recent Results makes perfect sense

However, as a QA guy, we really need a dashboard showing each product and the results of the running build, latest build, and ideally a few recent previous builds (but running+latest build are the minimal required)

This is the only sane way for QA Engineers/Release Managers to see the results they care about - how is their product and it's latest builds?

https://progress.opensuse.org/issues/6372 seems to be on the right track for addressing that need

So, to sum up my thoughts on this PR

As an activity overview/dashboard +1 - I wouldn't call it a 'Test Results' page any more, as it doesn't really focus on the 'results', but the general Activity of the openQA system.

Therefore generally speaking, we need a 'Build/Products Results' overview/dashboard also

**Comment by marita22** [comment]
I think for the Release Managers and me as QA Project Manager there is a need to not only have the build number but also the architecture.

1. Build numbers:
QA does acceptance tests for a specific build (= milestone candidate), once that is done, this special build will become a milestone that is released, e.g. for Partners, beta-program-participants, etc., it happens quite often that we get not only one candidate or a milestone but a second with some shipstopper fixes, for example

2. Architectures:
For Acceptance tests we have to look at each architecture separately and check if the candidate is ok. It happens quite often that some arches are ok and others are not, so we have to inform the Project Manager of that special arch. So an overview of testcases related to special arches is highly appreciated. But it is ok to go into a special build and get there an overview of testresults listed per architecture as it is implemented in the current UI.

3. Dashboard:
I like Coolos dashboard proposal very much but I miss the arches here:
e.g 
> sle-11-SP4-Alpha-Server-DVD
> Build0890: 21F 50P 1I 1O

Does that mean 21 Failures over all arches? It might be that we have got 7 tests failed on all three existing arches, it might be that we have different Testcases failed on different arches. Maybe it is possible to implement  a table with test results split per arch?

**Comment by coolo** [comment]
These 21 failures would link to e.g.
  https://openqa.suse.de/tests/overview?distri=sle&version=11-SP4-Alpha&build=0890

where you have all the details. The dashboard is just a way to collect these links with a summary. If we split this summary too much, we loose the adventage.

But with https://progress.opensuse.org/issues/6372 you would be able to couple or decouple products for the dashboard.

File: public/javascripts/tests.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		  var name = 'Build' + row['build'];
+		  name += " of ";
+		  return name + row['distri'] + "-" + row['flavor'] + "." + row['arch'];
+	      },
+	    },
+	    { targets: 1,
+	      className: "test",
+	      "render": function ( data, type, row ) {
+		  if (type === 'display') {
+		      html = '<a class="overview_' + row['result'] + '" href="tests/' + row['id'] + '">' + data + '</a>';
```
Comment (aaannz) [inline public/javascripts/tests.js]: I think there is missing leading / in the href URL and it creates double tests (i.e. http://localhost:9526/tests/tests/115)

File: public/javascripts/tests.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		      if (row['clone']) {
+                          html += ' <a href="/tests/' + row['clone'] + '">(restarted)</a>';
+                      } else {
+			  var url = restart_url.replace('REPLACEIT', row['id']);
+                          html += ' <a data-remote="true" href="' + url + '">' +
```
Comment (aaannz) [inline public/javascripts/tests.js]: This creates GET request, but jobs/restart is POST

File: templates/test/scheduled_table.html.ep [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		    % my $text = "Build$build of $settings->{DISTRI}-$settings->{VERSION}-$settings->{FLAVOR}.$settings->{ARCH}";
+		    %= $text
+                </td>
+		<td class="test">
+		    %= link_to $test => url_for('test', 'testid' => $job->id)
+		</td>
+                <td class="parents">
+		    % my $parents = $job->parents;
+		    % while (my $parent_id = $parents->next) {
+                        <a href="<%= url_for('test', 'testid' => $parent_id) %>">#<%= $parent_id %></a>
```
Comment (aaannz) [inline templates/test/scheduled_table.html.ep]: There should be $parent_id->id to get actual testid

File: public/javascripts/tests.js (hunk 1 above)
Comment (coolo) [inline public/javascripts/tests.js]: Hmm, I'm quite sure I checked the links, but you're right.

File: public/javascripts/tests.js (hunk 2 above)
Comment (coolo) [inline public/javascripts/tests.js]: data-remote is different, but I'll add a test to verify

File: templates/test/scheduled_table.html.ep (hunk 3 above)
Comment (coolo) [inline templates/test/scheduled_table.html.ep]: another missing test obviously - thanks for reviewing

File: public/javascripts/tests.js (hunk 1 above)
Comment (coolo) [inline public/javascripts/tests.js]: the relative links are mapped to absolute ones in phantomjs - for whatever reason. So this bug is actually not testable - I added the tests nonetheless and fixed the links

File: public/javascripts/tests.js (hunk 2 above)
Comment (coolo) [inline public/javascripts/tests.js]: you're right here too - and I noticed another problem while writing the test: the icon shouldn't be displayed unless logged in

File: templates/test/scheduled_table.html.ep (hunk 3 above)
Comment (coolo) [inline templates/test/scheduled_table.html.ep]: we have no test with parents in the fixtures ;(

File: lib/OpenQA/Controller/Admin/Table.pm [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Copyright (C) 2014 SUSE Linux Products GmbH
```
Comment (aaannz) [inline lib/OpenQA/Controller/Admin/Table.pm]: Wrong year and company, isn't it? And btw. I prefer GPL header without FSF address, but URL instead, once they move again it'll create tens of rpmlint warnings. And is recommended by https://www.gnu.org/licenses/gpl-howto.en.html

File: lib/OpenQA/Plugin/Helpers.pm [hunk 5]
```diff
@@ -21,6 +21,7 @@ use warnings;
 use Mojo::ByteStream;
 use db_helpers;
 use Data::Dumper;
+use Carp;
```
Comment (aaannz) [inline lib/OpenQA/Plugin/Helpers.pm]: nitpick, are DD and Carp needed?

File: lib/OpenQA/Scheduler.pm [hunk 6]
```diff
@@ -56,12 +56,12 @@ our (@ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS);
 @ISA = qw(Exporter);
 
 @EXPORT = qw(worker_register worker_get workers_get_dead_worker list_workers job_create
-  job_get job_get_by_workerid jobs_get_dead_worker list_jobs job_grab job_set_done
+  job_get job_get_by_workerid jobs_get_dead_worker list_jobs
```
Comment (aaannz) [inline lib/OpenQA/Scheduler.pm]: list_jobs is replaced by query_jobs

File: lib/OpenQA/Schema/Result/JobModules.pm [hunk 7]
```diff
@@ -109,14 +110,15 @@ sub job_module($$) {
...
 
     my $schema = OpenQA::Scheduler::schema();
-    return $schema->resultset("JobModules")->search({ job_id => $job->{id}, name => $name })->first;
+    return $schema->resultset("JobModules")->search({ job_id => $job->id, name => $name })->first;
 }
 
 sub job_modules($) {
     my ($job) = @_;
 
+    confess "JOB " if ref($job) eq 'HASH';
```
Comment (aaannz) [inline lib/OpenQA/Schema/Result/JobModules.pm]: That's leftover from testing I guess. And then the same D:D and Carp

File: lib/OpenQA/Utils.pm [hunk 8]
```diff
@@ -29,6 +26,7 @@ $VERSION = sprintf "%d.%03d", q$Revision: 1.12 $ =~ /(\d+)/g;
   &get_failed_needles
   &sanitize_testname
```
Comment (aaannz) [inline lib/OpenQA/Utils.pm]: sanitize_testname was also removed

File: lib/OpenQA/Controller/Admin/Table.pm (hunk 4 above)
Comment (coolo) [inline lib/OpenQA/Controller/Admin/Table.pm]: I just moved the file, but it was indeed very likely written in 2015 - ask Vlada :)

File: lib/OpenQA/Plugin/Helpers.pm (hunk 5 above)
Comment (coolo) [inline lib/OpenQA/Plugin/Helpers.pm]: I kind of like to have them around.

File: lib/OpenQA/Scheduler.pm (hunk 6 above)
Comment (coolo) [inline lib/OpenQA/Scheduler.pm]: Hmm, I wonder what this @EXPORT is for anyway - we use Scheduler::list_jobs anyway. But list_jobs is still used in api/v1/job - and we better don't touch the json export to avoid API breakage.

**Comment by aaannz** [comment]
I couldn't find it quickly enough but two more inconveniences:
- deps in scheduled jobs are not printed as links, only job#.
- the view is not dynamically updated (expect when restarting 'restarted' appears). So i.e. changing prio is visible only after reload, and even though 'restarted' appear after clicking, the scheduled list is updated only after page reload.

File: lib/OpenQA/Scheduler.pm (hunk 6 above)
Comment (aaannz) [inline lib/OpenQA/Scheduler.pm]: Ah sorry, I missed list_jobs was moved lower, I thought it was removed.
(aka 'I swear it wasn't there before!' )

**Comment by aaannz** [comment]
Another details:
- when the test starts and page is refreshed, Test time prints 'NaN years ago'
- when I didn't use buildnumber (I had a typo and used - instead =) in Scheduled and Running tables it prints 'Build of opensuse' and in Done tables it prints 'Buildnull of opensuse'
- Admin/Workers and JobTemplates pages are not yet converted to jQuery tables I guess, look different.

File: lib/OpenQA/Plugin/Helpers.pm (hunk 5 above)
Comment (aaannz) [inline lib/OpenQA/Plugin/Helpers.pm]: Funny thing I was removing them when present in whatever file I edited. Ok ok, I'll stop it now. :)

**Comment by coolo** [comment]
can you please file all these as "concrete bugs" in sprint 15? Most of them affect master too actually

**Comment by coolo** [comment]
removed dumper and debug output
